runtime/resetting_histogram.go

```diff
@@ -0,0 +1,305 @@
+func (t *StandardResettingHistogram) Snapshot() Histogram {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+	currentValues := t.values
+	t.values = make([]int64, 0, InitialResettingHistogramSliceCap)
+
+	sort.Sort(Int64Slice(currentValues))
+
+	return &ResettingHistogramSnapshot{
+		values: currentValues,
+	}
+}
+
+func (t *StandardResettingHistogram) Clear() {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+	t.values = make([]int64, 0, InitialResettingHistogramSliceCap)
+}
+
+// Record the duration of an event.
+func (t *StandardResettingHistogram) Update(d int64) {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+	t.values = append(t.values, d)
+}
+
+// ResettingHistogramSnapshot is a point-in-time copy of another ResettingHistogram.
+type ResettingHistogramSnapshot struct {
+	values              []int64
+	mean                float64
+	thresholdBoundaries []float64
+	calculated          bool
+}
+
+// Snapshot returns the snapshot.
+func (t *ResettingHistogramSnapshot) Snapshot() Histogram { return t }
+
+func (*ResettingHistogramSnapshot) Update(int64) {
+	panic("Update called on a ResettingHistogramSnapshot")
+}
+
+func (t *ResettingHistogramSnapshot) Clear() {
+	panic("Clear called on a ResettingHistogramSnapshot")
+}
+
+func (t *ResettingHistogramSnapshot) Sample() Sample {
+	panic("Sample called on a ResettingHistogramSnapshot")
+}
+
+func (t *ResettingHistogramSnapshot) Count() int64 { return int64(len(t.values)) }
+
+// Values returns all values from snapshot.
+func (t *ResettingHistogramSnapshot) Values() []int64 {
+	return t.values
+}
+
+func (t *ResettingHistogramSnapshot) Min() int64 {
+	if len(t.values) > 0 {
+		return t.values[0]
+	}
+	return 0
+}
+
+func (t *ResettingHistogramSnapshot) Variance() float64 {
+	if 0 == len(t.values) {
+		return 0.0
+	}
+	m := t.Mean()
+	var sum float64
+	for _, v := range t.values {
+		d := float64(v) - m
+		sum += d * d
+	}
+	return sum / float64(len(t.values))
+}
+
+func (t *ResettingHistogramSnapshot) Max() int64 {
+	if len(t.values) > 0 {
+		return t.values[len(t.values)-1]
+	}
+	return 0
+}
+
+func (t *ResettingHistogramSnapshot) StdDev() float64 {
+	return math.Sqrt(t.Variance())
+}
+
+func (t *ResettingHistogramSnapshot) Sum() int64 {
+	var sum int64
+	for _, v := range t.values {
+		sum += v
+	}
+	return sum
+}
+
+// Percentile returns the boundaries for the input percentiles.
+func (t *ResettingHistogramSnapshot) Percentile(percentile float64) float64 {
+	t.calc([]float64{percentile})
+
+	return t.thresholdBoundaries[0]
+}
+
+// Percentiles returns the boundaries for the input percentiles.
+func (t *ResettingHistogramSnapshot) Percentiles(percentiles []float64) []float64 {
+	t.calc(percentiles)
+
+	return t.thresholdBoundaries
+}
+
+// Mean returns the mean of the snapshotted values
+func (t *ResettingHistogramSnapshot) Mean() float64 {
+	if !t.calculated {
+		t.calc([]float64{})
+	}
+
+	return t.mean
+}
+
+func (t *ResettingHistogramSnapshot) calc(percentiles []float64) {
```

Comment:
What stops this from being a regular go-metrics snapshot?

Reply:
Not sure what you mean with that. The `calc` function is different, and the backend data structure is different, so we can't use the `HistogramSnapshot`.

Maybe it'd be possible to push this to the `Sample` type, but this is not really a sample, and we are only moving the logic one-level further.

Reply:
@coryschwartz we can't operate a go-metrics `HistogramSnapshot` because its state is not public: https://pkg.go.dev/github.com/rcrowley/go-metrics?tab=doc#HistogramSnapshot.